### PR TITLE
pwm_input modifications

### PR DESCRIPTION
**Pull request by arbulgazar**
** Reset feature added for pwm_input driver. **
GPIO_GPIO5_OUTPUT (AUX 6 signal pin) is currently used as reset pin.
Reset is happening if pwm_input don't get any data for some time, to prevent reset-spamming while lidar is not connected ROMFS should toogle pwm_input driver on parameter. (not implemented in this fork)

** Publication to range_finder topic added **
range_finder topic is used by several sensors (pwm_input, i2c lidar interface and range_finder currently) and ROMFS should have cases to enable ony one of them.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Thanks! Doing a full review now.

**Comment by LorenzMeier** [comment]
@Zefz This brings the Lidar-Lite to life. I need to think a moment about how to layer this architecturally.

**Comment by TSC21** [comment]
@LorenzMeier I was planning to get all rangefinders and altimeters data into `DISTANCE_SENSOR` message. Are there other plans (like keep using `range_finder` topic) or can I proceed with this? Mavros already is capable of sub and pub into `DISTANCE_SENSOR` msg.

File: makefiles/config_px4fmu-v2_default.mk [hunk 1]
```diff
@@ -70,6 +71,7 @@ MODULES		+= modules/commander
 MODULES		+= modules/navigator
 MODULES		+= modules/mavlink
 MODULES		+= modules/gpio_led
+MODULES         += modules/gpio_tool
```
Comment (arbulgazar) [inline makefiles/config_px4fmu-v2_default.mk]: This shouldn't be there. Included for testing, forgot to remove. :(

**Comment by Zefz** [comment]
LidarLite still needs working PWM drivers (#1949).

**Comment by TSC21** [comment]
@arbulgazar it would be awesome if you could do your changes against https://github.com/PX4/Firmware/pull/2200 :+1:

**Comment by bansiesta** [comment]
I can't get this to run yet, I keep getting:

```
nsh> pwm_input start
[pwm_input] advertising 536948912
nsh> [pwm_input] Lidar is down, reseting
[pwm_input] Lidar is down, reseting
[pwm_input] Lidar is down, reseting
...
```

Pin 5 and 6 are connected as shown here http://copter.ardupilot.com/wiki/common-optional-hardware/common-rangefinder-landingpage/common-rangefinder-lidarlite/#connecting_to_pixhawk_via_pwm, the rail is powered with 5v, and I'm using a 300 Ohm pull-down.

**Comment by LorenzMeier** [comment]
Try one kilo ohm instead

**Comment by LorenzMeier** [comment]
Actually I think it's rather software related. Did you put FMU in PWM mode and / or disabled the driver? I need to check if there are missing dependencies / startup order things.

**Comment by bansiesta** [comment]
Already re-soldering :). I didn't touch the fmu command...

**Comment by bansiesta** [comment]
I switched to 1k Ohm and disabled the fmu command, still nothing!
Does the Lidar-lite have any indication that it's running? Any idea how to debug this?

**Comment by TSC21** [comment]
@bansiesta are you using this https://github.com/PX4/Firmware/pull/1949?

**Comment by bansiesta** [comment]
@TSC21: I'm currently trying pwm_input. Is this not needed for ll40ls start pwm?

**Comment by TSC21** [comment]
As @Zefz said:
> LidarLite still needs working PWM drivers (#1949).

**Comment by bansiesta** [comment]
Sorry, don't get it. Tell me what I need to run in what order and what you suspect to work.

**Comment by TSC21** [comment]
@bansiesta I can't really tell. From what I know, I2C drivers are working for LidarLite but have problems related to bus hold, which is a problem related to LidarLite only. That's why a PWM driver is needed. But, from what I can see, there's no current implementation for using PWM with LidarLite in the firmware, and that's what @Zefz was implementing, but it didn't finished it. So, I suspect @Zefz code must be done first before getting this to work.

**Comment by bansiesta** [comment]
@TSC21 Understood thanks. I've already continued where @Zefz left off and I think I have an idea what is going on there. However, for the PWM part to work, this pwm_input needs to work but doesn't for me. The interrupt seems never triggered.

**Comment by bansiesta** [comment]
Aha, seems to work now, a ```pwm_input test``` was necessary.

**Comment by TSC21** [comment]
Good catch :D BTW, check https://github.com/PX4/Firmware/commit/43668cc8c8f6413861277aeab3fa026564b5c4fa. Big change on structure, as `range_finder` topic is totally replaced by `distance_sensor`.

**Comment by bansiesta** [comment]
@TSC21: thanks for the heads-up. I'll try to merge that in later.

**Comment by arbulgazar** [comment]
I wasn't modifying start-up scripts, was not sure you want lidar starting with fmu, but probably should have wrote this in comment:
`pwm_input start full` required in order to both initialize and start timer. `pwm_input start` will lead only to timer initialization. This however is listed in help message you get on `pwm_input`.

**Comment by arbulgazar** [comment]
I will do changes with respect to #2200 after week or so. Thesis is demanding time now, can't push it back.

**Comment by bansiesta** [comment]
@arbulgazar: Thanks for your work, I based my work on it and went from there. It's merged now.

**Btw:**
I removed the ```pwm_input start full``` argument because the read is now triggered from the ll40ls driver. Also the pwm_input only publishes to the pwm_input topic and the ll40ls driver then converts this to a distance, etc..
This should leave the pwm_input driver as a generic interface that can be extended for other use cases.

**Pull request merged**